src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Utility.cs

```diff
@@ -94,613 +71,275 @@ private static bool FirstCharInUInt32IsAscii(uint value)
-                    }
-
-                    pBuffer += 2 * SizeOfVector128;
-                } while (pBuffer <= pFinalVectorReadPos);
-            }
-
-            // We have somewhere between 0 and (2 * vector length) - 1 bytes remaining to read from.
-            // Since the above loop doesn't update bufferLength, we can't rely on its absolute value.
-            // But we _can_ rely on it to tell us how much remaining data must be drained by looking
-            // at what bits of it are set. This works because had we updated it within the loop above,
-            // we would've been adding 2 * SizeOfVector128 on each iteration, but we only care about
-            // bits which are less significant than those that the addition would've acted on.
+                    offset = 0;
+                    found = first;
+                    goto Found256;
+                }
+                if (Test256(last))
+                {
+                    offset = bufferLength - 32;
+                    found = last;
+                    goto Found256;
+                }
 
-            // If there is fewer than one vector length remaining, skip the next aligned read.
+                return bufferLength;
 
-            if ((bufferLength & SizeOfVector128) == 0)
-            {
-                goto DoFinalUnalignedVectorRead;
+            Found256:
+                found = Vector256.GreaterThanOrEqual(found, Vector256.Create((byte)0x80));
+                return ComputeIndex(found) + offset;
             }
 
-            // At least one full vector's worth of data remains, so we can safely read it.
-            // Remember, at this point pBuffer is still aligned.
+            Vector512<byte> found512;
+            ref byte pCurrent = ref pBuffer;
+            ref byte pLast = ref Unsafe.Add(ref pBuffer, bufferLength - 64);
 
-            if (Sse2.IsSupported)
+            // SAFETY: We only use the value of .AsPointer to calculate the byte offset
+            // to the next 64B boundary. Should GC relocate the buffer, we will lose the
+            // alignment which is an acceptable trade-off over pinning the buffer, once
+            // the callers of this method are updated to use byrefs.
+            nuint toAlign = (nuint)Unsafe.AsPointer(ref pCurrent) % 64;
+            if (toAlign != 0)
             {
-                currentSseMask = (uint)Sse2.MoveMask(Sse2.LoadAlignedVector128(pBuffer));
-                if (ContainsNonAsciiByte_Sse2(currentSseMask))
+                Vector512<byte> first512 = Vector512.LoadUnsafe(ref pCurrent);
+                if (Test512(first512))
                 {
-                    goto FoundNonAsciiDataInCurrentChunk;
+                    found512 = first512;
+                    goto Found512;
                 }
+
+                pCurrent = ref Unsafe.Add(ref pCurrent, toAlign);
             }
-            else if (AdvSimd.Arm64.IsSupported)
+
+            while (!Unsafe.IsAddressGreaterThan(ref pCurrent, ref pLast))
```

Comment:
As an early bit of feedback, this isn't going to pass review.

Manipulating the `byref` is dangerous and risks accidental GC holes and so we've take a stance to avoid that wherever possible. Instead you should be using the `Vector.LoadUnsafe(ref source, elementOffset)` overload to help ensure things remain correct.

Reply:
`pCurrent` here is always within bounds, but it does require care. Will change it to offset-based indexing.

Reply:
> pCurrent here is always within bounds, but it does require care

Yes, it looks like this code is correct, but its also something that has been gotten wrong in the past and the perf wins don't outweigh the risk, so we avoid such code where possible.

CC. @jkotas, @AndyAyersMS 

> Will change it to offset-based indexing.

👍